imp_developer.py

```diff
@@ -1305,55 +2044,217 @@ def __init__(self, env):
+        self.state = self.IDLE
+    #
+    def check_imp_error(self, error):
+        if not error:
+            return False
+
+        # Handle invalid credentials use-case only
+        # the following string should restart logs via command
+        # which lead to access token renew
+        if error["code"] == ImpRequest.INVALID_CREDENTIALS:
+            sublime.set_timeout_async(
+                lambda: self.env.window.run_command("imp_show_console", {"cmd_on_complete": "auth"}), 0)
+
+        return True
+
+
+    def __read_logs(self):
+        if self.sock and type(self.sock) != None and self.sock.fp != None:
+            next_log = False
+            next_cmd = False
+            logs = []
+            count = PL_LOGS_MAX_PER_REQUEST
+            while count > 0:
+                # lets read no more than coun logs per one loop
+                rd, wd, ed = select.select([self.sock], [], [], 0)
+                if not rd:
+                    break
+                else:
+                    for line in self.sock:
+                        if next_log:
+                            next_log = False
+                            logs.append(line.decode("utf-8"))
+                        elif next_cmd:
+                            next_cmd = False
+                            if line == b'data: closed\n':
+                                self.reset()
+                                logs.append("Stream was closed by server event.\n")
+                                return logs
+                            if line == b'data: opened\n':
+                                self.keep_alive = datetime.datetime.now()
+                                return logs
+                            if line != b'\n':
+                                logs.append(line.decode("utf-8"))
+
+                        elif line == b'event: message\n':
+                            count -= 1
+                            next_log = True
+
+                        elif line == b'event: state_change\n':
+                            count -= 1
+                            next_cmd = True
+                        elif line == b'\n':
+                            next_cmd = False
+                            next_log = False
+                            break
+                        elif line == b': keep-alive\n':
+                            self.keep_alive = datetime.datetime.now()
+                        else:
+                            log_debug("Unhandled command:" + str(line.decode("utf-8")))
+
+            # check that stream was not dropped
+            if self.keep_alive != None and len(logs) == 0:
+                delta = datetime.datetime.now() - self.keep_alive
+                if delta.seconds > PL_KEEP_ALIVE_TIMEOUT:
+                    log_debug("Did not get keep alive on-time. Trigger log reset.")
+                    self.reset()
+
+            # work around to keep stream active when logs are available
+            if len(logs) > 0:
```

Comment:
let's please move this block before the previous one

Reply:
fixed as a part of the previous comment